File: System-Backend/scripts/update-tailscale.py
#!/usr/bin/env python3
"""
Tailscale IP Auto-Update Script (Backend)
Automatically fetches the current Tailscale IPv4 address and updates config.py
"""

import subprocess
import sys
import re
from pathlib import Path

def get_tailscale_ip():
    """Fetch current Tailscale IPv4 address"""
    try:
        result = subprocess.run(
            ['tailscale', 'ip', '-4'],
            capture_output=True,
            text=True,
            timeout=5,
            check=True
        )
        ip = result.stdout.strip()
        
        # Validate IP format (must be 100.x.x.x)
        if not re.match(r'^100\.\d+\.\d+\.\d+$', ip):
            raise ValueError(f"Invalid Tailscale IP format: {ip}")
        
        return ip
    except subprocess.CalledProcessError:
        print("❌ Error: Tailscale command failed")
        print("   Make sure Tailscale is running: tailscale status")
        sys.exit(1)
    except FileNotFoundError:
        print("❌ Error: Tailscale is not installed or not in PATH")
        print("   Install Tailscale: https://tailscale.com/download")
        sys.exit(1)
    except subprocess.TimeoutExpired:
        print("❌ Error: Tailscale command timed out")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Error: {e}")
        sys.exit(1)

def update_config(tailscale_ip):
    """Update config.py with the current Tailscale IP"""
    config_file = Path(__file__).parent.parent / 'config.py'
    
    if not config_file.exists():
        print(f"❌ Error: config.py not found at {config_file}")
        sys.exit(1)
    
    try:
        content = config_file.read_text()
        
        # Update TAILSCALE_IP default value
        updated_content = re.sub(
            r'TAILSCALE_IP = os\.getenv\("TAILSCALE_IP", "[^"]*"\)',
            f'TAILSCALE_IP = os.getenv("TAILSCALE_IP", "{tailscale_ip}")',
            content
        )
        
        # Update OLLAMA_HOST default value to use Tailscale IP
        updated_content = re.sub(
            r'OLLAMA_HOST = os\.getenv\("OLLAMA_HOST", f?"http://[^"]*:11434"\)',
            f'OLLAMA_HOST = os.getenv("OLLAMA_HOST", f"http://{tailscale_ip}:11434")',
            updated_content
        )
        
        config_file.write_text(updated_content)
        print(f"✅ Backend Tailscale IP updated successfully!")
        print(f"   IP: {tailscale_ip}")
        print(f"   File: config.py")
        print(f"   OLLAMA_ENDPOINT: http://{tailscale_ip}:11434/api/generate")
        return True
    except Exception as e:
        print(f"❌ Error updating config.py: {e}")
        sys.exit(1)

if __name__ == "__main__":
    print("🔍 [BACKEND] Fetching Tailscale IPv4 address...")
    tailscale_ip = get_tailscale_ip()
    update_config(tailscale_ip)

File: System-Frontend/scripts/update-tailscale.js
#!/usr/bin/env node
/**
 * Tailscale IP Auto-Update Script
 * Automatically fetches the current Tailscale IPv4 address and updates constants/env.ts
 * 
 * Usage: node scripts/update-tailscale.js
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const ENV_FILE = path.join(__dirname, '../constants/env.ts');

try {
  // Get current Tailscale IP
  console.log('🔍 Fetching Tailscale IPv4 address...');
  const tailscaleIP = execSync('tailscale ip -4', {
    encoding: 'utf-8',
    timeout: 5000,
  }).trim();

  if (!tailscaleIP || !tailscaleIP.match(/^100\.\d+\.\d+\.\d+$/)) {
    throw new Error(`Invalid Tailscale IP format: ${tailscaleIP}`);
  }

  // Read current env.ts
  const envContent = fs.readFileSync(ENV_FILE, 'utf-8');

  // Replace the IP placeholder with actual IP
  const updatedContent = envContent.replace(
    /export const TAILSCALE_IP = '[^']*'/,
    `export const TAILSCALE_IP = '${tailscaleIP}'`
  );

  // Write back to file
  fs.writeFileSync(ENV_FILE, updatedContent, 'utf-8');

  console.log('✅ Tailscale IP updated successfully!');
  console.log(`   IP: ${tailscaleIP}`);
  console.log(`   File: constants/env.ts`);
  console.log(`   API_BASE_URL: http://${tailscaleIP}:8000`);
} catch (error) {
  if (error.status === 127) {
    console.error('❌ Error: Tailscale is not installed or not in PATH');
    console.error('   Install Tailscale: https://tailscale.com/download');
    process.exit(1);
  } else if (error.message.includes('timeout')) {
    console.error('❌ Error: Tailscale command timed out');
    console.error('   Make sure Tailscale is running: tailscale status');
    process.exit(1);
  } else {
    console.error('❌ Error updating Tailscale IP:', error.message);
    process.exit(1);
  }
}
